internal/yaml.go

```diff
@@ -0,0 +1,157 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/maps"
+)
+
+// Stats represent the expected/actual amount of
+// failed, skipped and passed tests.
+type Stats struct {
+	UnexpectedRest int `yaml:"unexpected_rest"`
+	UnexpectedFail int `yaml:"unexpected_fail"`
+	UnexpectedSkip int `yaml:"unexpected_skip"`
+	UnexpectedPass int `yaml:"unexpected_pass"`
+	ExpectedFail   int `yaml:"expected_fail"`
+	ExpectedSkip   int `yaml:"expected_skip"`
+	ExpectedPass   int `yaml:"expected_pass"`
+}
+
+// ConfigYAML is a yaml tests representation of the Config struct.
+//
+// It is used only to fetch data from file. To get any of
+// the dance configuration data it should be converted to
+// Config struct with Convert() function.
+//
+//nolint:govet // we don't care about alignment there
+type ConfigYAML struct { // ConfigYAML -> Config
+	Runner  string      `yaml:"runner"`
+	Dir     string      `yaml:"dir"`
+	Args    []string    `yaml:"args"`
+	Results ResultsYAML `yaml:"results"`
+}
+
+// ResultsYAML is a yaml representation of the Results struct.
+type ResultsYAML struct { // ResultsYAML -> Results
+	Common   *TestsConfigYAML `yaml:"common"`
+	FerretDB *TestsConfigYAML `yaml:"ferretdb"`
+	MongoDB  *TestsConfigYAML `yaml:"mongodb"`
+}
+
+// TestsConfigYAML is a yaml representation of the TestsConfig struct.
+// It differs from it by using "any" type to be able to parse maps (i.e. "- output_regex: ...").
+//
+// To gain a data the struct should be first converted to TestsConfig with TestsConfigYAML.Convert() function.
+type TestsConfigYAML struct { // TestsConfigYAML -> TestsConfig
+	Default status `yaml:"default"`
+	Stats   *Stats `yaml:"stats"`
+	Pass    []any  `yaml:"pass"`
+	Skip    []any  `yaml:"skip"`
+	Fail    []any  `yaml:"fail"`
+}
+
+// Convert converts ConfigYAML to the Config struct.
+func (cf *ConfigYAML) Convert() (*Config, error) {
+	common, err := cf.Results.Common.Convert()
+	if err != nil {
+		return nil, err
+	}
+	ferretDB, err := cf.Results.FerretDB.Convert()
+	if err != nil {
+		return nil, err
+	}
+	mongoDB, err := cf.Results.MongoDB.Convert()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Config{
+		cf.Runner,
+		cf.Dir,
+		cf.Args,
+		Results{common, ferretDB, mongoDB},
+	}, nil
+}
+
+// Convert converts a TestsConfigYAML to the TestsConfig struct.
+// TestsConfigYAML is a yaml file with the tests.
+// TestsConfig is an internal representation of yaml test file.
```

Comment:
> // TestsConfig is an internal representation of yaml test file.

The comment is for `TestsConfigYAML`. Let's start the comment with `TestsConfigYAML` word or with `Convert` word. Otherwise, I wonder if this is a comment for this function or if I was just copied from somewhere). 


```suggestion
// Convert validates yaml and converts TestsConfigYAML to the internal representation of yaml test file, TestsConfig.
```